sled-agent/src/nexus.rs

```diff
@@ -202,3 +167,816 @@ impl ConvertInto<nexus_client::types::DatasetKind>
+    pub total_put_requests_completed: u64,
+    pub cancelled_pending_notifications: u64,
+}
+
+#[derive(Debug)]
+pub struct NexusNotifierHandle {
+    tx: mpsc::Sender<NexusNotifierMsg>,
+}
+
+#[derive(Debug)]
+pub struct SenderOrReceiverDropped {}
+
+impl NexusNotifierHandle {
+    pub async fn notify_nexus_about_self(&self, log: &Logger) {
+        if let Err(_) =
+            self.tx.send(NexusNotifierMsg::NotifyNexusAboutSelf).await
+        {
+            warn!(log, "Failed to send to NexusNotifierTask: did it exit?");
+        }
+    }
+
+    #[allow(unused)]
+    pub async fn get_status(
+        &self,
+    ) -> Result<NexusNotifierTaskStatus, SenderOrReceiverDropped> {
+        let (tx, rx) = oneshot::channel();
+        self.tx
+            .send(NexusNotifierMsg::Status(tx))
+            .await
+            .map_err(|_| SenderOrReceiverDropped {})?;
+        rx.await.map_err(|_| SenderOrReceiverDropped {})
+    }
+}
+
+// The type of operation issued to nexus
+enum NexusOp {
+    Get,
+    Put,
+}
+
+/// What sled-agent has confirmed that Nexus knows about this sled-agent
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum NexusKnownInfo {
+    // CRDB doesn't contain a record for this sled-agent
+    NotFound,
+    Found(SledAgentInfo),
+}
+
+impl NexusKnownInfo {
+    fn generation(&self) -> Generation {
+        match self {
+            NexusKnownInfo::NotFound => Generation::new(),
+            NexusKnownInfo::Found(known) => known.generation,
+        }
+    }
+}
+
+/// Return the latest sled-agent info. Boxed for use in testing.
+type GetSledAgentInfo = Box<dyn Fn(Generation) -> SledAgentInfo + Send>;
+
+// A mechanism owned by the `NexusNotifierTask` that allows it to access
+// enough information to send a `SledAgentInfo` to Nexus.
+pub struct NexusNotifierInput {
+    pub sled_id: Uuid,
+    pub sled_address: SocketAddrV6,
+    pub nexus_client: NexusClient,
+    pub hardware: HardwareManager,
+    pub vmm_reservoir_manager: VmmReservoirManagerHandle,
+}
+
+// Type returned from a join on the "outstanding request" task`
+type NexusRsp = (
+    NexusOp,
+    Result<SledAgentInfo, nexus_client::Error<nexus_client::types::Error>>,
+);
+
+/// A mechanism for notifying nexus about this sled agent
+///
+/// The semantics are as follows:
+///  1. At any time there is a single outstanding HTTP request to nexus
+///  2. On startup, this task gets the latest sled-agent info, if any, from
+///     nexus, and saves it.
+///  3. Whenever the state needs to be updated to a value different
+///     from what nexus has, the generation number is bumped
+///     and the new state transmitted.
+///  4. If a caller requests an update to be made to nexus and it succeeds
+///     the known value is set to what was updated.
+///  5. If the request fails, we go ahead and set the known state to `None`,
+///     since we are not sure if the last update succeeded. This will
+///     trigger a get request for the latest state.
```

Comment:
It might be worth noting here that the request for the newest state is load-bearing, in the sense that an update can fail because the task's state is out-of-date with respect to what's in Nexus, and the only way to fix that problem is to get the newest Nexus state. (And it's simplest just to do that unconditionally than to try to reason about exactly why it was that Nexus rejected one of these requests.)